### PR TITLE
feat: improve the transfer flow for irc2 token

### PR DESCRIPTION
**Pull request by dust-leaf**

### CODE REVIEW

**Comment by parrot9design** [comment]
@bam-21, let's keep all related discussion to this PR for easier tracking.

For incomplete transfers:

- [ ] Allow users to input any address to complete the transfer.
- [ ] Update the UI message to allow users to remove the asset from the contract without needing an address:

>{amount} {asset} is awaiting transfer.<br>[Remove it from the bridge contract](#), or enter an address to complete the transaction.
slice1@2x

- [ ] Update the red used for the error text to `#fb6a6a`

----

- [ ] Update the banner message displayed when adding/removing assets like sICX to/from the contract pre-transfer.

Current:
>Approving {asset} for cross-chain transfers...
>Approved {asset} for cross-chain transfers. 
>Removing {asset} from the contract...
>Removed {asset} from the contract.

Preferred:
>Sending {asset} to the bridge contract...
>Sent {asset} to the bridge contract.
>Removing {asset} from the bridge contract...
>Removed {asset} from the bridge contract.

----

- [ ] Fix the link text here – the underline should only appear on hover.
Screenshot 2023-07-25 at 11 24 50 AM

----

- [ ] I'm being prompted to send to contract / complete abandoned transactions for ICX (from BSC->ICON) in this deployment, when that's not the case in the live app. IIRC, this flow is only needed for IRC2 tokens like sICX, bnUSD, BALN, etc. ?

**Comment by luanvuonggia** [comment]
After the asset has been sent to the contract, display a Remove from contract text link instead of check icon
Screenshot 2023-07-27 at 14 04 13

**Comment by bam-21** [comment]
- [ ] In case of insufficient funds, the modal is displaying minus number of tokens.
image

**Comment by dadengit** [comment]
To be consistent, I see the link "Remove it from bridge contract" should be the same as the "Learn more" link, underline when hover. @parrot9design @bam-21

**Comment by bam-21** [comment]
@dadengit I agree.

**Comment by parrot9design** [comment]
>To be consistent, I see the link "Remove it from bridge contract" should be the same as the "Learn more" link, underline when hover.

Yep, let's add an underline to the link text on hover – nice catch!

---

> In case of insufficient funds, the modal is displaying minus number of tokens. Update to `{amount} {asset} is required to transfer`, with the amount = transfer amount + transfer fee

I say we either a) update the message to `You need at least {amount} {asset} to transfer cross-chain`, or b) remove it completely and let the `Insufficient {token}` button label take care of it. I'm leaning towards b.

**Comment by parrot9design** [comment]
To better facilitate testing, let's keep this branch pointed to mainnet instead of Berlin testnet.

**Comment by trieuvo212** [comment]
@mangcanbang @parrot9design It looks good to me

**Comment by bam-21** [comment]
@mangcanbang @parrot9design it looks good to me.

**Comment by parrot9design** [comment]
- [ ] This is a confirmation message, so it only needs one `.` at the end instead of `...`:
Screenshot 2023-08-07 at 11 31 36 AM

---

- [ ] `Remove it from the bridge contract` should remain green on hover:
Screenshot 2023-08-07 at 11 31 55 AM

---

- [ ] Let's fix the second line to say `Removing {asset} from the bridge contract...` and use the regular text colour instead of green as it's no longer a link:
Screenshot 2023-08-07 at 11 32 11 AM

---

The messages have gotten a bit mixed up when transferring from BSC to ICON. The initial transaction through MetaMask is to approve the spending cap, not send assets to a contract. Yet while I see the `Approve` button when I go to transfer an asset, everything after that in the modal and banner notifications talks about the bridge contract instead. The current version on app.balanced.network handles this better.
Screenshot 2023-08-07 at 11 42 57 AM

**Comment by mangcanbang** [comment]
I fixed all your comments @parrot9design , thank you very much for your review!

**Comment by parrot9design** [comment]
Thanks @mangcanbang! 

Let's change the `Approve` button label to `Approve asset`, like it is the current UI.

After an asset has been approved for cross-chain transfers via Metamask, the modal still displays the `Remove from contract` link and the subsequent banner messages talk about removing the asset from the contract even though you're actually revoking the spending cap.

Screenshot 2023-08-11 at 9 41 35 AM
<br><br>

We should either display the option to revoke the spending cap set in Metamask (`Revoke asset`), or the tick symbol used in the current UI:
Screenshot 2023-08-11 at 9 43 58 AM

If you cancel out of the confirmation modal after approving the asset, you have to approve the asset again before you can complete the transfer, even if you don't change any of the details. Should this be the case?

**Comment by mangcanbang** [comment]
Pic 1. After approve asset
Screenshot 2023-08-13 at 23 47 11
Pic 2. Approving asset
Screenshot 2023-08-13 at 23 47 24
Pic 3. Approve success, the transfer modal show remove from contract link
Screenshot 2023-08-13 at 23 47 40
Pic 4. Removing 
Screenshot 2023-08-13 at 23 47 52

Pic 5. Remove success, the screen still shows the transfer modal for the user can approve this asset again.
Screenshot 2023-08-13 at 23 47 59

@parrot9design You want to show the tick symbol in which picture?

**Comment by parrot9design** [comment]
Pics 3 & 4. In this scenario the asset has been _approved_ for transfer – it's not held in a contract so there's nothing to remove.

We should either replace `Remove from contract` with the tick, or, if we want to give people a direct way to revoke their spending limit in MetaMask, we should use `Revoke asset` and these banner messages:

>Revoking cross-chain transfers for {asset}...

>Revoked cross-chain transfers for {asset}.

**Comment by mangcanbang** [comment]
I change the label off approve button and replace `Remove from contract ` with the tick @parrot9design

**Comment by parrot9design** [comment]
@mangcanbang can you point the branch to mainnet? It reverted back to Berlin where it's difficult to test.

**Comment by mangcanbang** [comment]
@parrot9design IDK why it still shows Berlin, but actually it is mainnet, help me connect your wallet and test again.

**Comment by hetfly** [comment]
It's set to mainnet now

@mangcanbang it's done through Vercel, you can set environment variables there for different builds

**Review by parrot9design — APPROVED** [review]
Looks good to go from a UX perspective. Thanks so much for humouring all my requests, @mangcanbang. 🙏

@hetfly are you wanting to give this a code review before we merge?

File: .gitignore
Comment (hetfly) [inline .gitignore:31]: keep .env in gitignore, vercel is used for handling that.

File: src/btp/src/connectors/MetaMask/index.ts
Comment (hetfly) [inline src/btp/src/connectors/MetaMask/index.ts:86]: If this was for the testing only, don't forget to revert it

File: src/btp/src/connectors/chainConfigs.js
Comment (hetfly) [inline src/btp/src/connectors/chainConfigs.js:90]: same here

**Review by hetfly — CHANGES_REQUESTED** [review]

**Comment by hetfly** [comment]
Other than the above, looks good to me

File: .gitignore
Comment (mangcanbang) [reply .gitignore]: Okay thank you, I will practice it.

**Pull request merged**
